src/game.cpp

```diff
@@ -7968,10 +7984,122 @@ bool pet_menu(monster *z)
 
         return true;
     }
+    if ( inject == choice ) {
+        auto filter = [] (const item &it) {
+            for (auto cont : it.contents) {
+                return cont.has_flag( "MUTAGEN_CEPHALOPOD" ) || cont.has_flag( "PURIFIER" );
+            }
+            return it.has_flag( "MUTAGEN_CEPHALOPOD" ) || it.has_flag( "PURIFIER" );
+        };
 
-    return true;
-}
+        int pos = g->inv_for_filter( _("Injectable mutagen:"), filter );
+        if ( pos == INT_MIN ) {
+            add_msg( _("Never mind.") );
+            return true;
+        }
+        item &it = g->u.i_at(pos);
+
+        item &injectable = it;
+
+        bool container = false;
 
+        for (auto &cont : it.contents) {
+            if (cont.has_flag( "MUTAGEN_CEPHALOPOD" ) ||
+                cont.has_flag( "PURIFIER" ) ) {
+                injectable = cont;
```

Comment:
This will not work properly, because here you're overwriting `injectable` with `cont`. References can't be reassigned like that, instead it invokes `item::operator=` here.

You could change `injectable` to a pointer or (better option) write a small helper function that returns first content that has a given flag (or a flag from a set of flags).

You don't really need to support the case where the player selected mutagen directly, without a container. Such cases don't happen in game. Only contained mutagen needs to work.